### ISSUE TITLE
[RFC] NestedTensor API - 0.0.1

### ISSUE THREAD

**Issue by cpuhrsch** [pytorch/pytorch#22169]
## Motivation
Many fields manipulate collections of Tensors of different shapes. For example, paragraphs of text, images of different sizes or audio files of different lengths. We don't have a first class abstraction that enables the concurrent manipulation of collections of this type of data. We further often need to batch arbitrary data and operations for efficiency.

## Definition
A NestedTensor, just like a torch.Tensor, has a dtype, layout, device and dimension.

In general, there are two cases for which NestedTensors provide computational representations.

**Lists of Tensors** 
Each Tensor constituent of the list it represents, if any, must be of its dtype, layout and device. The dimension of a constituent Tensor must be one less than the dimension of the NestedTensor. An empty list of Tensors yields a NestedTensor of dimension one.

**Lists of NestedTensors**
Each constituent NestedTensor must be of its dtype, layout and device. The dimension of a constituent NestedTensor must be one less than the dimension of the NestedTensor. 

You cannot have a list of Tensors and NestedTensors, that is, you cannot mix them. 

## Goals and current code

**Convenience:** Remove tedious manual workarounds

Sometimes research can be unblocked by or inspired by convenient abstractions. New code may be built on top of NestedTensors if it's general enough and easy enough to use (discoverable). We want to make sure that we solve existing problems well, but also want it to be general enough to inspire new work.

```
# This does a lot of work to use our awkward packed sequence abstraction
def forward(self, sent_tuple):
    # sent_len: [len_1, ..., len_bsize]
    # sent: (seqlen x bsize x worddim)
    sent, sent_len = sent_tuple

    # Sort by length (keep idx)
    sent_len_sorted, idx_sort = np.sort(sent_len)[::-1], np.argsort(-sent_len)
    sent_len_sorted = sent_len_sorted.copy()
    idx_unsort = np.argsort(idx_sort)

    idx_sort = torch.from_numpy(idx_sort).cuda() if self.is_cuda() \
        else torch.from_numpy(idx_sort)
    sent = sent.index_select(1, idx_sort)

    # Handling padding in Recurrent Networks
    sent_packed = nn.utils.rnn.pack_padded_sequence(sent, sent_len_sorted)
    sent_output = self.enc_lstm(sent_packed)[0]  # seqlen x batch x 2*nhid
    sent_output = nn.utils.rnn.pad_packed_sequence(sent_output)[0]

    # Un-sort by length
    idx_unsort = torch.from_numpy(idx_unsort).cuda() if self.is_cuda() \
        else torch.from_numpy(idx_unsort)
    sent_output = sent_output.index_select(1, idx_unsort)
```

**Correctness:** Abstract the subtle details

Many users run into subtle issues when padding and masking manually. We can hide these details and provide an implementation that deals with all these subtleties. 

Example 1
```
# Currently users have to worry about setting their masked values to
# the correct number to not influence reductions such as max.
def neginf(dtype):
    """Returns a representable finite number near -inf for a dtype."""
    if dtype is torch.float16:
        return -NEAR_INF_FP16
    else:
        return -NEAR_INF
        
[...]
outputs_of_interest = embedding_layer[:, 1:, :]
# The user carries around a mask to indicate regions that aren't relevant
# This mask needs to be updated for each layer.
mask = (~attention_mask[:, 1:]).type(dtype).unsqueeze(2) * neginf(dtype)
sumed_embeddings = torch.sum(outputs_of_interest * mask, dim=1)
nb_elems = torch.sum(attention_mask[:, 1:].float(), dim=1).unsqueeze(1)
embeddings = sumed_embeddings / nb_elems
# This is a very complicated way of getting torch.mean(outputs_of_interest, dim=1)
# for variably sized entries.
[...]
```

Example 2
```
tensor *= mask.unsqueeze(-1).float()
for i in range(self.n_layers):
    # The user manually carries around a mask to indicate regions
    # that aren't relevant. Each layer has custom mask support
    # which the user had to implement
    tensor = self.layers[i](tensor, mask)

# It's very easy to write -1e20 instead of 1e-20, which makes a big difference
divisor = mask.float().sum(dim=1).unsqueeze(-1).clamp(min=1e-20)
output = tensor.sum(dim=1) / divisor
```

**Performance:** Take advantage of sparsity and ease batching

There are obvious gains in writing specialized kernels. RNNs already take advantage of that via packed sequences, which was introduce by a vendor library. We can take advantage of this more broadly via a centralized abstraction. It also allows us to replace for loops or other obviously inefficient constructs.

```
GeneralizedRCNNTransform.forward(self, images, targets=None):
    # If the images are small this likely causes resource underutilization.
    for i in range(len(images)):
        image = images[i]
        target = targets[i] if [...]
        [...]
        image = self.normalize(image)
        image, target = self.resize(image, target)
        images[i] = image
        if targets is not None:
            targets[i] = target
    image_sizes = [img.shape[-2:] for img in images]
    images = self.batch_images(images)
    image_list = ImageList(images, image_sizes)
    return image_list, targets
```

## Scope of this document
For this first version we're treating NestedTensors effectively as a static container (akin to Python tuples) of its constituent Tensors. That means, any operation for explicit shape manipulation, such as broadcasting or select, are not part of this discussion and will be brought up as part of 0.0.2.

## Python repr() string representation
A NestedTensor of two one-dimensional Tensors of length three and four.
```
nestedtensor([
 tensor([8, 1, 3, 4]),
 tensor([5, 0, 9])
]) 
```

A NestedTensor of NestedTensors each containing Tensors of different lengths.
```
nestedtensor([
 nestedtensor([
  tensor([8, 1, 3, 4]),
  tensor([5, 0, 9])
 ]),
nestedtensor([
  tensor([2, 4]),
  tensor([5, 0, 9, 10, 2])
 ])
])
```
 
## Construction/Conversion
For this discussion we use the Python frontend for notation.

### Construction
```
$ a = [torch.rand(2, 3), torch.rand(4, 5)]
$ b = torch.nestedtensor(a)
```

The constructor also accepts tuples.
```
$ a = (torch.nestedtensor([torch.rand(2, 3)]),
       torch.nestedtensor([torch.rand(4, 5)]))
$ b = torch.nestedtensor(a)
```

The level of nesting is inferred from the input. The constructor always copies. Whatever you pass into the constructor will share no data with what the constructor returns. This matches torch.tensor's behavior.

If given a NestedTensor or Tensor it will return a detached copy, which is consistent with the behavior of torch.tensor. Remember that you cannot mix Tensors and NestedTensors within a given list.

### Conversion/unbind()
Converting from NestedTensors to nested tuples
```
$ a = [ \
       [torch.rand(2, 3), torch.rand(4, 5)], \
       [torch.rand(6, 7)] \
      ]
$ b = torch.nestedtensor(a)
$ b1 = b.unbind() # Tuple of 2 NestedTensors
$ b2 = b1[0].unbind() # Tuple of 2 Tensors
```

A user can retrieve the constituent Tensors via unbind. Unbind is currently used by torch to turn Tensors into tuples of Tensors. Unbind always returns a tuple of views. We do not (yet support a dimension argument to unbind for NestedTensors, because it forces us to argue about shape.

## Operator generalization
Operator generalization is the process of expanding input constraints and semantics to NestedTensors. Note that this also includes features such as aliasing (creating views). Every Tensor operator, if meaningfully generalizable, may be implemented or must be flagged as not implemented and raise an exception.

We directly add support for NestedTensors to operations in torch and there will be no specialized package.

As a general policy, we will be conservative in the operations we support and introduce them slowly for this more general case. In some cases the advanced semantics are obvious, in others they will be heavily driven by potential applications and current issues.

### Generalization strategy
We can find that many operations of torch.Tensor can be expressed as a combination of others. We can minimize the amount of work we need to do in order to get good operator coverage by identifying core operations and generalizing them. This will then also act as a forcing function to yield sane and consistent semantics, because the combination of these core operations will also imply combined semantics and input constraints.

### nested_size()
As motivation let us consider as an example of two paragraphs of word ids.
```
$ words = \
[
 [
  tensor([0, 2, 4, 6]), 
  tensor([1, 3, 5])
 ],
 [
  tensor([7, 9]), 
  tensor([11]), 
  tensor([13, 15, 17, 19]),
 ] 
]
$ words = torch.nestedtensor(words)
$ words.nested_size()
(
 (
  torch.Size([4]),
  torch.Size([3])
 ), 
 (
  torch.Size([2]),
  torch.Size([1]),
  torch.Size([4])
 )
)
```
nested_size() returns a container of torch.Sizes just like NestedTensor is a container of torch.Tensors. We could also introduce a torch.NestedSize for debatable benefit, but increased consistency.

Note: Someone might call a dimension ragged or batched if for a given dim `size(dim)` doesn't match across all Tensor constituents. We say that this dimension is heterogeneous.

### nested_dim()
```
$ words.nested_dim()
(
 (1, 1),
 (1, 1, 1)
)
```

Similar to nested_size, nested_dim returns a nested representation. Due to the constraints on the constituent Tensors all numbers will be equal, but this might change in the future if we relax some of them.

### Definition

Let's define `z = f(Tensor x_1, Tensor x_2,..., Tensor x_n)` to be an operation on n Tensors. Let's say data is a list of n valid inputs to the constructor.
```
# data is a list that represents inputs to our constructors
# Example:
# data = [(torch.rand(2, 3), torch.rand(5, 4)), (torch.rand(7, 9),)]
inputs = [torch.nestedtensor(data_i) for data_i in data]

def tw_f(inputs, f):
  result = []
  if isinstance(inputs[0], torch.Tensor):
    return f(*inputs)
  else:
    result = []
    # Turn this list of NestedTensors into a list of NestedTensor constituents
    unbound_inputs = [input.unbind() for input in inputs]
    for i in range(len(inputs[0])):
      result.append(tw_f([unbound_input[i] for unbound_input in unbound_inputs], f))
    return torch.nestedtensor(result)
    
result = tw_f(inputs, f)
```
We call an operator *tw(f) to be tensor-wise with respect to f* if it is, semantically, implemented via above code. 

**We impose two input constraints for this operation (for now).**

1. The resulting list of Tensors or NestedTensors (called result in the above snippet) need to form a valid NestedTensor. This typically requires that all n input NestedTensors be of the same dtype, layout and dimension for that construction to succeed. We make this a strict requirement.
2 Each NestedTensor constructed from data_i must have the same nesting structure. This means nested_dim must match across all n input NestedTensors.

### Example: torch.mm
```
$ a = torch.nestedtensor(
[
 [
  rand(3, 5)
 ],
 [
  rand(1, 8),
  rand(7, 3)
 ]
])

$ b = torch.nestedtensor(
[
 [
  rand(5, 3)
 ],
 [
  rand(8, 2),
  rand(3, 4)
 ]
])

$ torch.matmul(a, b).nested_size()
(
 (
  torch.Size([3, 3]),
 ),
 (
  torch.Size([1, 2]),
  torch.Size([7, 4])
 )
)
```

We can think of the NestedTensor version of mm as the tensor-wise operation based on mm.

**List of supported tensor-wise operations**

What follows is a list of categories of operations we consider tensor-wise. All of these operations may  be implemented as part of 0.0.1. Some of these stand in place for Python class methods (such as __eq__) and may also double for a module level function, such as torch.add, and method function, such as torch.Tensor.add_.

1. Unary point-wise operations such as add, cos and digamma
2. Binary operations such as add and cmul, but limited to working without Tensor broadcasting
2. mm, mv, addmm and other BLAS operations
4. spectral operations such as fft and ifft
5. specific nn layers such as Embedding/EmbeddingBag or RNN

If there is any doubt whether we want to treat these functions as tensor-wise in the future, we should not implement them as such for now.

Note that we explicitly punt on broadcasting semantics within this RFC simply because they can be viewed as implicit shape manipulation of a NestedTensor (even if it's only of a Tensor constituent) and we want to discuss that separately.

As set of heuristics for inclusion in this list is
1. The operator is independent of the sizes of the Tensor inputs (e.g. torch.cos)
2. The operator does not care about the sizes of the Tensor inputs as long as they match (e.g. torch.add)
3. The operator currently does not have batching or broadcasting semantics (e.g. torch.mm or torch.fft)
4. The operator currently has ad-hoc or awkward variable length input support that can be replaced by NestedTensors

### Embedding
```
def embedding_bag(
    input,
    weight,
    offsets=None,
    max_norm=None, norm_type=2,
    scale_grad_by_freq=False, mode='mean',
    sparse=False, per_sample_weights=None)
```

Currently, if input is of size (B, N) it will be treated as B sequences of length N. The offsets are ignored. The function returns a Tensor of size (B, em), where em is the embedding dimension. If input is 1d, then offsets is used to slice up the input vector. The offsets are currently used to emulate variable length sequences.

Using NestedTensor we can simply do away with offsets. The input is a 2d NestedTensor of size ((n_1,), (n_2,), ..., (n_k,)) and the output will be of size ((em,), (em,), ..., (em,)). This is the tensor-wise operation based on EmbeddingBag applied to a (1, N) Tensor. If there are B Tensor constituents in the input NestedTensor we will simply return a torch.Tensor of size (B, em).

### RNNs
A NestedTensor of 1-dimensional Tensors can be converted into a PackedSequence. RNN functions accept PackedSequences. As such, we can treat NestedTensors as a replacement for PackedSequence with the additional benefits of supporting empty Tensor constituents and entries not sorted by length. PackedSequence is another example of ad-hoc support for variable length sequences, however via a more sophisticated, separate class.

### cross_entropy
```
cross_entropy(
    input,
    target,
    weight=None,
    size_average=None,
    ignore_index=-100,
    reduce=None,
    reduction='mean')
```
cross_entropy has implicit support for variable length data via the ignore_index flag. We can implement support for variable length inputs explicitly via NestedTensors, if we treat it as a tensor-wise operation based on cross_entropy for a single sequence.

### Lists of Tensors: Data tensor plus mask tensor

We may start out with a dense tensor and a mask tensor. For performance concerns we may raise a warning on low load factors.

**Advantages**
- All pointwise n-ary operations are immediately implemented by applying them to the data tensor.
- Many torch.nn operations are implemented in torch methods and functions, which yields immediate coverage.
- Many operations can be implemented in terms of others, e.g. BLAS in terms of matmul and index_select via masked_select.
- Autograd support comes out of the box, since this can be written was a wrapper.
- In many cases an existing user implementation uses padding plus masking, so a replacement doesn't yield a performance regression, but surely makes it easier for the user and more widely available.
- Any dtype, layout and device is immediately supported.
- Vendor libraries that require torch.Tensor for high performance operations such as matrix multiply still can be used.
- Provides a central location for various existing implementations that use padding and masking.

**Disadvantages**
- Doesn't take advantage of sparsity and adds a lot of memory overhead due to the mask.
- Cannot support a relaxation of the dtype, layout or device constraints, because there has to be one underlying data Tensor.

### List of NestedTensors

Lists of NestedTensors effectively add metadata on top of a NestedTensor of Tensors. As such they can be implemented separately or even concurrently with future work. We may also restrict ourselves to only a few levels of nesting at first to ease implementation.

### Lists of Dense Tensors: Packed data layout

Eventually we will want to use a packed data layout to benefit from sparsity. For this we discuss two cases.

1. A NestedTensor of Tensors
2. A NestedTensor of NestedTensors

We will restrict ourselves to NestedTensors with dense layout, regular C dtype and allocated on a device with pointer arithmetic (e.g. CPU or CUDA GPU).

![image](http://blog.ezyang.com/img/pytorch-internals/slide-08.png)

 See above for a visual refresher on our memory layout for dense Tensors. For more information and credit for this picture go to http://blog.ezyang.com/2019/05/pytorch-internals/.

**Case 1:**
Let's consider a NestedTensor called t of the following form.
```
nestedtensor([
  tensor([...]),
  tensor([...]),
  ...
  tensor([...])
])
```
Let's say there are K Tensors and let's call the kth Tensor in this sequence t_k. We can characterize its size by `(n_k^1, n_k^2, ..., n_k^N)` and strides `(s_k^1, s_k^2, ..., s_k^N)`. In concordance with torch.Tensor all strides are non-negative. Call the pointer to the underlying memory data_k. Each logical datapoint maps to a datapoint in the corresponding memory, but two logical data points may map to the same if e.g. the strides are 0.

For each logical element of t_k there exists a unique sequences of indices `(i_k^1, i_k^2, ..., i_k^N)` that maps to the underlying memory location via `data_k + \sum_{n = 1}^{N} i_k^n * s_k^n`. The maximum offset is `s_k^0 = \sum_{i = 1}^{N} i_k^n * s_k^n`.

Pick one logical element x of t. WLOG we can assume x is a part of t_k. Then we can map to the underlying data via `data_k + \sum_{n = 1}^{N} i_k^n * s_k^n` and associate index `(k, i_k^1, i_k^2, ..., i_k^N)`.

Now ,to have a packed data representation we can, during construction or other times, union the memory of each data_k into a single contiguous memory location pointed to by data. Then we can define a mapping from each logical element with index `(k, i_k^1, i_k^2, ..., i_k^N)` via `data + \sum_{i = 0}^{k - 1} s_i^0 + \sum_{n = 1}^{N} i_k^n * s_k^n` to the underlying data.

In conclusion, we simply unify the underlying data into a single memory region. We only need K further values via `s_0^k` to store the offsets and implement this memory layout. This also makes it much easier to provide views and adopt current implementations for regular Tensors. An initial implementation might not unify the data and simply stores a vector of Tensors with metadata to represent the nesting on top.

**Case 2:**
Consider a NestedTensor of the following form.
```
nestedtensor([
  nestedtensor([...]),
  nestedtensor([...]),
  ...
  nestedtensor([...])
])
```
Let's say there are K NestedTensors and let's call the kth NestedTensor in this sequence nt_k. We can then proceed just as we did with our Tensors and introduce a new index i_k^0 with corresponding offset s_k^0 that equals the maximum offset of each constituent. 

**A note on metadata**
However, note that adding another level of nesting due to additional NestedTensors does not change the amount of underlying allocated memory. It simply increases the amount of metadata which is used to represent the structure on top of the collection of Tensors. We might, in the future, want to combat this by introducing homogeneity annotations per dimension (e.g., nested_size() must return the same number for each torch.Size at a given dimension). This will allow us to store a single stride instead of one stride per constituent and might also make it easier to leverage existing kernels. We can also use this to compress the metadata we need to store for the Tensor constituents.

**Advantages**
- Due to our homogeneity constraint on dtype, device and layout much of the current code should be easy to extend. In particular, TensorIterators should be easy to extend for pointwise operations with additional striding information (vector of vector of strides)
- Might concurrently resolve discussions around blocked memory supports (vector of vector of strides)

**Disadvantages**
- Requires deep integration and possibly extension of core structures
- We can run into issues with allocation overhead if we store this as a std::vector of Tensors.
- Some vendor library operations require regular torch.Tensors. High performance operations such as matrix multiplication might require new kernels, but this may be eased with homogeneity annotations.

## Next steps
As next steps we can implement a basic prototype located in torch.prototype.nestedtensor. It'll be a Python wrapper around two torch.Tensors, one data Tensor and one mask Tensor. It allows construction from a list of torch.Tensors, i.e. no nesting. Dispatch is implemented via monkey-patching torch on import of torch.prototype.nestedtensor.torch and using isinstance to distinguish between torch.Tensor and torch.NestedTensor. It also comes with a test suite within test/test_nestedtensor.py.

**Comment by mcarilli** [comment]
It seems like this proposal is basically equivalent to a convenience wrapper that ingests a list of Tensors (or list of lists, whatever), does a deep-copy flatten of the list, then creates a similarly-structured list of views into the flat buffer, and exposes that list of views as the NestedTensor.

Advantages that I see:
- In principle, you can rig the above approach using only Python logic, without changing Aten.
- Since the constructor always copies, you can ensure the Tensors owned by NestedTensor are C-style contiguous.
- The above approach makes pointwise ops on the entire NestedTensor (or a sub-NestedTensor) easy:  just fire off a single pointwise kernel on the flat buffer.  You can directly use existing Aten ops, and they'll even be autogradable, as a bonus.
- I don't think you need to store an auxiliary mask.  At worst you should insert a bit of padding as you flatten to ensure that each owned Tensor's storage starts on a memory-aligned (say 128-byte aligned) boundary.  The padding would not require too much additional memory, I don't think (at least no more than allocating the equivalent set of scattered tensors, since the scattered allocations are aligned anyway).

-----
The "multi tensor apply" functionality in Apex is used somewhat differently.  Every time I want to perform a multi-tensor operation, I create (on the fly) a set of shallow references to an existing set of non-flattened tensors at different locations in memory, then launch a single backend op that processes all of them in a single or a few kernel launches.  The intention is to enable optimizers for which different sets of gradients may be present/created at each iteration (e.g. some of the gradients may be None).  I construct the list of params to process/update at each iteration based on which params have a gradient.

Advantages of Apex-style approach:
- It does not require explicitly flattening anything.  It can operate on scattered tensors where they lie.
- Building the list of tensors on the fly allows ops to easily and adaptively change what tensors are going to be operated on in each iteration (for example) without deep memory movement.

Is there a reason you chose to prefer creating NestedTensors via a deep copy instead of allowing them to be a list of views?  The list-of-views approach introduces complications, but many of those complications we have already handled.  I'm wondering why you made this decision and if there's some roadblocks I haven't considered.


-----
In general, I think optimizers should be considered alongside RNNs/embedding/cross entropy as a first class example application.  Optimizers are a ubiquitous case where identical pointwise ops are applied to many Tensors in a loop.  For typical networks, this results in hundreds of small kernel launches, which often becomes bottlenecked by CPU-side overhead.  We've found that optimizers benefit significantly from parameter flattening or a backend op that can process many Tensors in a small number of kernel launches.

**Comment by cpuhrsch** [comment]
> It seems like this proposal is basically equivalent to a convenience wrapper that ingests a list of Tensors (or list of lists, whatever), does a deep-copy flatten of the list, then creates a similarly-structured list of views into the flat buffer, and exposes that list of views as the NestedTensor.
> 
> Advantages that I see:
> 
> * In principle, you can rig the above approach using only Python logic, without changing Aten.
> * Since the constructor always copies, you can ensure the Tensors owned by NestedTensor are C-style contiguous.
> * The above approach makes pointwise ops on the entire NestedTensor (or a sub-NestedTensor) easy:  just fire off a single pointwise kernel on the flat buffer.  You can directly use existing Aten ops, and they'll even be autogradable, as a bonus.
> * I don't think you need to store an auxiliary mask.  At worst you should insert a bit of padding as you flatten to ensure that each owned Tensor's storage starts on a memory-aligned (say 128-byte aligned) boundary.  The padding would not require too much additional memory, I don't think (at least no more than allocating the equivalent set of scattered tensors, since the scattered allocations are aligned anyway).
I would not store an auxiliary mask for the packed data layout, it's not required. Indeed you can think of this as a single flat-buffer and the list of Tensors as a list of views. However, they don't have to be contiguous. Each Tensor can have different strides and sizes. They also don't *have* to share memory, we could, as an extension, make that a trait similar to "is_contiguous" ("is_packed"?), which then, as you mentioned, can make it easier to run certain pointwise operations. More so, if each Tensor constituent has the same last dimension and the flat buffer is "packed" we might even launch kernels that represent more complex operations, since we can treat the flat buffer as an array of vectors (with possibly some padding inbetween).
> 
> The "multi tensor apply" functionality in Apex is used somewhat differently. Every time I want to perform a multi-tensor operation, I create (on the fly) a set of shallow references to an existing set of non-flattened tensors at different locations in memory, then launch a single backend op that processes all of them in a single or a few kernel launches. The intention is to enable optimizers for which different sets of gradients may be present/created at each iteration (e.g. some of the gradients may be None). I construct the list of params to process/update at each iteration based on which params have a gradient.

I don't see why you couldn't do this with a NestedTensor of packed data layout. You can treat each Tensor independently and, at most, they happen to share memory via a single flat-buffer. However, gradients are indeed more complicated in that case.
> 
> Advantages of Apex-style approach:
> 
> * It does not require explicitly flattening anything.  It can operate on scattered tensors where they lie.
> * Building the list of tensors on the fly allows ops to easily and adaptively change what tensors are going to be operated on in each iteration (for example) without deep memory movement.
This is still the case here, no? The packed data layout should be able to provide an interface similar to a std::vector<at::Tensor>. Also, NestedTensors are user facing constructs. You can still, at the backend, convert it into this list of tensors on the fly without incurring any cost, right?
> 
> Is there a reason you chose to prefer creating NestedTensors via a deep copy instead of allowing them to be a list of views? The list-of-views approach introduces complications, but many of those complications we have already handled. I'm wondering why you made this decision and if there's some roadblocks I haven't considered.
This is consistent with the behavior of torch.tensor. Creating a NestedTensor that shares memory with a list of Tensors should be implemented via a separate function and might restrict it in various ways. 
> 
> In general, I think optimizers should be considered alongside RNNs/embedding/cross entropy as a first class example application. Optimizers are a ubiquitous case where identical pointwise ops are applied to many Tensors in a loop. For typical networks, this results in hundreds of small kernel launches, which often becomes bottlenecked by CPU-side overhead. We've found that optimizers benefit significantly from parameter flattening or a backend op that can process many Tensors in a small number of kernel launches.
I'm happy to include an example here. Do you have a particular suggestion / code snippet?

**Comment by mcarilli** [comment]
> I don't see why you couldn't do this with a NestedTensor of packed data layout. You can treat each Tensor independently and, at most, they happen to share memory via a single flat-buffer. However, gradients are indeed more complicated in that case.

Of course it's doable by "treating each tensor independently" but I'm not sure how treating each member of a NestedTensor "independently" would be different from processing non-nested Tensors in a loop (which is functionally correct, but not performant).  The performance gain with multi tensor apply comes from treating the tensors homogeneously and processing them all with a single launch.

> I'm happy to include an example here. Do you have a particular suggestion / code snippet?

The inner loop of any optimizer is a clear example, e.g. https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L88-L106

**Comment by cpuhrsch** [comment]
> Of course it's doable by "treating each tensor independently" but I'm not sure how treating each member of a NestedTensor "independently" would be different from processing non-nested Tensors in a loop (which is functionally correct, but not performant). The performance gain with multi tensor apply comes from treating the tensors homogeneously and processing them all with a single launch.

So your main concern is that a NestedTensor cannot be created from a list of Tensors without a deep copy? That's not a strict requirement, it's simply mirroring the current behavior of torch.tensor.  But there can be other functions or constructors that do what you.

> The inner loop of any optimizer is a clear example, e.g. https://github.com/pytorch/pytorch/blob/master/torch/optim/sgd.py#L88-L106

Thanks

> In general, my wariness of the deep-copy requirement is because you may wish to launch one-off aggregate operations on a list of tensors that are freshly created each iteration (ie gradients, or data-batch elements, or any temporaries internal to the model). The deep copy is a high price to pay for such a one-off operation, and may negate the performance benefit of allowing the backend to aggregate the operation.

If you're creating a list of tensors at each iteration, why wouldn't that be a NestedTensor which represents a list of Tensors?

**Comment by mcarilli** [comment]
> If you're creating a list of tensors at each iteration, why wouldn't that be a NestedTensor which represents a list of Tensors?

The tensors might be created separately, in scattered locations.  At the point where the loop/common operation must be carried out, it would be nice to pack them into a NestedTensor as shallow views, then call operations on the NestedTensor, which can then dispatch to a backend that "vectorizes" over the owned Tensors efficiently.

**Comment by cpuhrsch** [comment]
> > If you're creating a list of tensors at each iteration, why wouldn't that be a NestedTensor which represents a list of Tensors?
> 
> The tensors might be created separately, in scattered locations. At the point where the loop/common operation must be carried out, it would be nice to pack them into a NestedTensor as shallow views, then call operations on the NestedTensor, which can then dispatch to a backend that "vectorizes" over the owned Tensors efficiently.

From a technical perspective it doesn't like anything is preventing us from doing this, if we introduce a separate "packed" annotation that allows applying operations on a list of Tensors with arbitrary memory location and if we introduce a function that can create a view. Maybe something like torch.nestedtensor_view?

Let's think about this more separately. I'll write it down as a requirement.

**Comment by mcarilli** [comment]
To be clear, by "pack them into a NestedTensor as shallow views," I meant explicitly NOT creating any new dense/packed storage for them.  By "packing" I meant simply stashing a list of references to the existing tensors where they lie.  The purpose of the NestedTensor wrapper would be to serve as a dispatcher to a backend capable of processing the list of scattered references efficiently.

**Comment by mcarilli** [comment]
@cpuhrsch How do you imagine autograd working for a NestedTensor?  Let's say you carry out
`n2 = n0*n1`, where n0 and n1 are NestedTensors whose owned tensors all require grad (or, even worse, maybe some of them require grad and some of them don't).  How would the autograd history be recorded in this case?  How would the gradients be passed back?

You said something about 
> We may start out with a dense tensor and a mask tensor. For performance concerns we may raise a warning on low load factors.
> ...
> Autograd support comes out of the box, since this can be written was a wrapper.

but I'm not entirely sure how this would work, unless the implementation of `operator*` simply acted on owned Tensors individually under the hood, which would eliminate the potential performance gain of being able to act on them in a batched way.

Making it a requirement that `T.requires_grad=True|False` must be the same for all Tensors within a given NestedTensor could simplify things.

**Comment by cpuhrsch** [comment]
@mcarilli - NestedTensor is still supposed to be a tensor, so the idea is that anything like requires_grad or detach acts on the Tensor as a whole and not individually. If you implement operator* with a data and masking Tensor for two given NestedTensors a and b and with result stored in c, you'd check that a, b and c's mask Tensors match and multiply a and b's data Tensors with the result stored in c. Or more concretely ```assert a.mask == b.mask; assert b.mask == c.mask; torch.mul(a.tensor, b.tensor, out=c.tensor)```. You'd then get autograd support from the fact that both .mask and .tensor are regular torch.Tensors with autograd support and torch.mul is a torch operation with autograd support. That's what I mean by "Autograd support comes out of the box, since this can be written has a wrapper.".

If you want a list of Tensors, we have a [TensorList](https://github.com/pytorch/pytorch/blob/65b00aa5972e23b2a70aa60dec5125671a3d7153/aten/src/ATen/core/Type.h#L39) via ArrayRef. There should also be nothing preventing you from using that to write ATen kernels and assume they are independent. We can then convert to TensorLists on the backend when appropriate, for example for particular NestedTensor operations, and use the associated kernels. I think this actually exactly aligns with what you're looking for?

**Comment by mcarilli** [comment]
Do you expect/require that the gradient created for a NestedTensor is also flattened/contiguously stored?  This could become a big problem for performance if people extract (unbind) individual Tensors as slices of the contiguously-allocated NestedTensor, perform autograd ops on the individual Tensors, then backward into these individual Tensors.  Currently, the backward implementation of slicing creates a gradient for _each_ small sliced-out Tensor that has the full size of the big buffer that Tensor was sliced from, and adds that full-sized gradient on top of any existing full-size gradient that the big buffer may already possess.

**Comment by cpuhrsch** [comment]
@mcarilli - No, we should make the packed layout optional and use it when it makes sense.

**Comment by mcarilli** [comment]
> EDIT: Let me dig deeper into the backwards behavior of slicing.

You'll find it's not ideal.  Each slice that is sliced from a large buffer will create a full-size gradient buffer and accumulate it into the [InputBuffer](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L572) for the large buffers `grad_fn`.

**Comment by cpuhrsch** [comment]
@mcarilli - I updated the issue to incorporate our discussion.

For NestedTensor we, in particular, need to worry about [unbind_backward](https://github.com/pytorch/pytorch/blob/7707dee76169a2c2d6637f80af5ed59eeb32a997/tools/autograd/templates/Functions.cpp#L436). Here we could easily add some optimizations.

Overall, do you agree that at::TensorList is enough for you to write kernels that can make optimal use of your code? You can pass in arbitrary lists of Tensors. NestedTensors might then use these kernels. For TensorList in particular we can invoke those functions by passing a list of Python functions from the Python frontend. However, one of the main concerns here is the allocation of [overhead of a lot of small Tensors](https://github.com/pytorch/pytorch/issues/13049). This is something we need to look at in any case however.

**Comment by mcarilli** [comment]
Why does unbinding need to be an autogradable operation at all?  Can't constituent Tensors just receive gradients depending on how they are individually or collectively used?

**Comment by cpuhrsch** [comment]
@mcarilli well you said

>This could become a big problem for performance if people extract (unbind) individual Tensors as slices of the contiguously-allocated NestedTensor, perform autograd ops on the individual Tensors, then backward into these individual Tensors.

**Comment by mcarilli** [comment]
Do you think it makes sense to store the owned Tensors individually in a non-flattened way, even when they're being deep-copy-constructed from the input tensors?  Then we won't have to worry about any slicing backwards.  If the user wants to retrieve an individual Tensor from the NestedTensor, we can just give them the reference.  And if the user wants to use this individual Tensor for autograd operations, it will work as normal, and receive its own gradient.  Neither gradients nor Tensors will be flattened in this case, but they will be more convenient to hand back individually.

**Comment by cpuhrsch** [comment]
@mcarilli - Yes, I think we should store them individually, but we can still have them share memory and keep track of that fact as an annotation or detect it at runtime based on the addresses of the pointers.

The main idea here is that you might have a collection of Tensors each with the same inner dimension. Then you can think of them as a single collection of Vectors for operations such as matrix multiplication.

**Comment by mcarilli** [comment]
Ah yeah that's true, if we store them contiguously, a batched gemm that operates on each of them becomes a single gemm launch.

> I think we should store them individually, but we can still have them share memory
what do you mean?  that seems contradictory

**Comment by cpuhrsch** [comment]
@mcarilli - we can have a collection of Tensors and they each view a piece of the same memory, or put differently, the memory of each Tensor "borders" with the memory of one other Tensor in the collection. So, when we allocator a Tensor, we allocate enough memory for all Tensors at once and then have them each point to some dedicated piece of memory.

**Comment by mcarilli** [comment]
I see, that is what I meant by "flattened" above.  I was wondering whether it made more sense to store them in a non-flattened way, but horizontal fusion of gemms is a compelling use case for storing them as views into a flattened buffer.

**Comment by cpuhrsch** [comment]
@mcarilli - can't we detect conditions that the horizontal fusion is appropriate at runtime? For example, if each Tensor is contiguous, each has the same inner dimension and the underlying pointers add up based on the maximum offset of each Tensor (i.e. we know that they collectively cover a dense, contiguous region of memory)? Unbinding operations etc. etc. *might* change this condition, but it is still something we can dispatch at runtime on.

**Comment by mcarilli** [comment]
Sure, but presumably we are making the decision about how to allocate and lay out the memory at the time when the NestedTensor is constructed, not at the time it's passed to a gemm, for example.  So the user would have to anticipate how it's going to be used, and request contiguous backing memory (or not).

**Comment by cpuhrsch** [comment]
@mcarilli - it's not quite unlike calling .contiguous() on a torch.Tensor. We can verify the condition at runtime (like we do with contiguity) and otherwise, just like we do now, have to call .contiguous() (and possibly suffer a reallocation) at various points of the program.

**Comment by mcarilli** [comment]
Well the downside is for one-off use cases, like in an optimizer.step(), where a bunch of gradients have been created at disjoint locations, and you want to apply them individually to a bunch of params.  In this instance the NestedTensor's role would be to replace a for loop over params+gradients by shallowly wrapping a list of params+gradients and dispatching to a backend capable of processing the entire set of params+gradients efficiently (via some multi tensor apply kernel).  A flattening deep copy in this instance is wasteful.  It would be nice if the params were already flattened all along, which isn't that difficult to arrange, but the gradients must also been created in a flattened way.**

**It's possible to pre-arrange that gradients will wind up in a flat buffer.  If `.grad` attributes are manually assigned as views into a flat buffer buffer before training, those locations will not move, and they will be accumulated into.  However, accumulating into an existing `.grad` attribute (even if it's just zeros) requires 3 passes over global memory (read incoming gradient, read `.grad`'s existing data, add and write to `.grad`).  That's why we have the `param.grad=None` optimization:  If a `.grad` attribute is None at the time its leaf receives a gradient, that leaf's AccumulateGrad function can simply steal the reference to the gradient handed back by the last autograd function,  and assign that reference directly to its `.grad` attribute, removing the need to accumulate and saving 3 passes over global memory.  Therefore, setting `p.grad=None` for all params before backward, instead of zeroing them via `model.zero_grad()` or `optimizer.zero_grad()` is an easy-to-use and often beneficial performance optimization.  The downside is that setting grads to None prevents you from predetermining where they will be created, because it leaves you at the mercy of the backward autograd op to create the gradient whose reference you steal to become your `.grad` attribute.  That backward autograd op is likely calling (for example) `empty_like` internally to create the gradient, which will create the gradient at a random location in memory, and you have know way to tell it to populate the gradient at some _desired_ location in memory, short of rewriting the op itself.
tl;dr It's possible to tell pytorch to send gradients to sections of a flat buffer, but doing so prevents use of the nifty grad=None optimization.

Another reason for wanting to shallow-wrap a list of potentially disjoint tensors and gradients on the fly in the optimizer is that some params may not have received gradients in this particular iteration.  Some of them may have None gradients. You only want to apply gradients to the params whose gradients are not None, so in my case I construct a list of the params whose gradients are not None before launching any multi tensor operations.

**Comment by mcarilli** [comment]
(transferring comments from our Slack conversation)

My idea is that `nested_tensor.contiguous()` (which may be called, optionally, as part of the NestedTensor constructor) should:

- Create the flat buffer Tensor via `torch.cat(owned_tensors)`.  This cat call _is_ exposed to autograd and recorded by autograd.
- `set_()` the storage for the owned tensors to be pointers into the newly catted buffer.

Whenever a collective operation is requested from the nested tensor, we carry it out on the post-cat flat buffer Tensor.  This way, autograd forward ops will be efficient, and gemms will be fused.  Backward ops that received the flat buffer in forward will accumulate into the cat backward's `InputBuffer` (which is a full-sized flat buffer) in the autograd engine.

Once all backward usages of the full buffer are resolved, cat backward will slice views of the accumulated big gradient buffer and hand one back to each member Tensor.

And if any member Tensors are unbound and used individually, they will receive gradients individually.  These individual-use gradients and the sliced-gradient views from the full-buffer/collective uses will be combined (accumulated) by the InputBuffer each member Tensor's `grad_fn`.

Benefits of this approach:
- Simple to implement.
- We don't need "unbind" to do anything special at all.  We can just give the user a reference to the owned Tensor, and the user can use it individually however they like.
- Fast, fused collective operations when requested
- Operations on/extraction of individual member tensors doesn't require slicing, therefore won't incur any nasty large-buffer creation in backward.
- grad=None optimization still works
- With this approach as-is, we don't _need_ custom autograd functions or new kernels for anything (not forward, not backward, not unbind).  The only custom autograd ops we would like to have (as performance optimizations) are:
  - Multi-tensor apply kernels (which I can provide) that enable fast collective operations on _non-contiguous_ NestedTensors.
  - A "repair cat history" operation to restore the autograd history of the flat buffer so it points to the individual Tensors (i'll shortly describe why/when this is helpful, as well as an easy implementation suggestion).

I've gotta catch up with my family now, I'll post more details this evening.

**Comment by mcarilli** [comment]
Some loose ends that I wanted to record more thoughts on:
- It turns out cat backward is implemented via `narrow` (see `cat_tensors_backward` in torch/csrc/autograd/generated/Functions.cpp).  This is good news for single-backward (it means that the proposal above does what we hoped it would and is optimal in terms of memory management as long as we stick to  single-backward), but if we run a double-backward, we will incur the nasty per-tensor large-buffer creation associated with `narrow`/slicing backward.
- It's definitely still useful to have an option where NestedTensor can be constructed as a thin holder of references to many Tensors that are scattered in memory (as opposed to deep copying them into a big flat buffer).  NestedTensor's role would be to serve as a dispatcher to some (future) multi-tensor collective backend (which for present API purposes could simply be a for loop).  This would serve use cases where you just want to shallowly wrap a bunch of Tensors and dispatch a fast collective op or two, without paying the price of a deep-copy contiguification.  `optimizer.step()` is my pet example of such a situation.
- It might be nice to pad tensors to, say, 128-byte alignment as we are catting them to ensure aligned access for later operations on individual member tensors.

**Comment by mcarilli** [comment]
Another random thought I just had:
```
a = Tensor(...)
b = Tensor(...)
nested = NestedTensor((a, b), contiguous=True) # or whatever syntax you want
```
For `contiguous=True`, you likely intend that `nested[0]` and `a` do NOT point to the same deep memory at this point, but they could be made to.  In the NestedTensor constructor, we could do either
```
torch.cat((a,b))
a.set_(...)
b.set_(...)
```
or
```
ac = a.clone()
bc = b.clone()
torch.cat((ac,bc))
ac.set_(...)
bc.set_(...)
```
Perhaps the constructor could have two orthogonal options:  `contiguous=True|False` and `deepcopy = True|False`.

**Comment by cpuhrsch** [comment]
@mcarilli - to answer a few of your questions:

Just like we have [torch.as_tensor](https://pytorch.org/docs/master/torch.html#torch.as_tensor) we should also have a torch.as_nestedtensor, which does not do a deep copy. However, it will still have the homogeneity constraints for its member variables. This includes that each owned tensor must have the same "requires_grad" status, same "pinned_memory" status etc. - If we then make them contiguous via a torch.cat and .set_ call we'll overwrite the memory. The individual torch.Tensor structs that share memory with the catted tensor then maybe be used for various operations or the catted tensor might be used for various operations. I don't understand, aside from read-write race conditions, I don't understand how we can run into issues with catted Tensor's backward aside from the issue we'd already run into. It'll also be used in operations that need to make sure of the entire buffer.

**Comment by mcarilli** [comment]
Homogeneity constraints sound like a good idea.

With the cat + set_ approach, single-backward is fine and basically optimal in terms of the amount of deep memory movement that occurs.  However, double-backward is still inefficient.  cat backward uses `narrow`, which means cat double-backward incurs per-tensor large buffer creation and accumulation.  This is not a correctness issue, it'll give the right answer, but performance will be slow.  Off the top of my head, I can't figure out an easy way to make double-backward as performant as single-backward, but I'm not sure we should regard it as blocker for now.  Double-backward is the less common case.

**Comment by cpuhrsch** [comment]
Unfortunately you [cannot use set_](https://pytorch.org/docs/master/tensors.html#torch.Tensor.detach) on a tracked Tensor, because it is an inplace operation. So, contiguous would only work if the owned Tensors do not require gradients (which they surely will want to do). Now, instead of doing cat + set_, I'm looking into a list version of clone (via a C++ extensions) that allocates memory for all given Tensors at once and then returns a cloned version, which is still tracked by autograd, where all the Tensor happen to have memory that "lives" next to each other.

Kernels that want to exploit contiguity will then need to check that all given Tensors are contiguous and "live next to each other", that is their offsets and data pointers add up. If that condition is fulfilled it is possible to create a few of the overall storage, but this is akin to TensorAccessors it will need to happen in C++ again.

**Comment by mcarilli** [comment]
The cat+set_ trick will work if the set_ calls are done under a `with torch.no_grad()` context.  Creating the flat buffer (the `cat` call) should still be exposed to autograd.

**Comment by mcarilli** [comment]
One thing I just realized is that the cat+set_ approach, if you choose to try it, only works (and should only be done) when constructing a NestedTensor from an initial set of incoming (non-flat) Tensors.

For operations where _existing_ NestedTensors are combined to create a new NestedTensor in an autogradable way, for example,
```
Nested3 = Nested1*Nested2
```
here's what you can try under the hood as a first pass:
```
Nested3.flat_buffer = Nested1.flat_buffer*Nested2.flat_buffer
Nested3.individual_member_views = torch.split(Nested3.flat_buffer, individual_member_sizes)
```
If any of Nested3's individual member Tensors are later extracted and used individually, split backward is not too bad in terms of the amount of deep memory creation/accumulation that occurs.  It will create a single full-sized buffer for all the gradients received by individual members (as opposed to a separate full-sized buffer being created for the gradient received by each individual member, which is what would happen if you created the individual member views by slicing instead of splitting).

I'm thinking more and more that for ideal performance we can define a custom autograd op that stitches together an efficiently-executable history for the flat buffer and the individual member views, which we can interpose either just before or just after the flat buffer is created (whether that creation is a constructor call or the result of an op like `N3 = N1*N2`).  I also think this custom autograd op can closely resemble split_with_sizes backward.  [split_with_sizes_backward is awesomely clever](https://github.com/pytorch/pytorch/blob/9130ab380a4af0c67036ea8fce937156cfd0a8c8/tools/autograd/templates/Functions.cpp#L754-L774)  and definitely worth looking at for inspiration.  split_with_sizes_backward feels like the right way to take advantage of the fact that all the splitted tensors are being created at once (as opposed to slicing in a loop, which paranoically treats each sliced-out view individually, and therefore in backward creates gradients for each sliced-out view that are the full size of the original big flat buffer).

**Comment by Gsunshine** [comment]
`torch.autograd.grad` is also a desirable tensor-wise op for NestedTensor 0.0.1. This is a toy implementtaion of bacth jacobian via `torch.autograd.grad` and python for-loop. However, jacobian is a tensor with aligned memory. Is NestedTensor memory-efficient for batched jacobian, especially when we do a double backpropagation over jacobian?

```
def batch_jacobian(y, x):
    '''
    Compute the Jacobian matrix in batch.
    Return (B, D_y, D_x)
    '''

    batch = y.shape[0]
    y = y.view(batch, -1)
    y_size = y.shape[1]
    vector = torch.ones(batch).to(y)

    # Compute Jacobian row by row.
    # dy_i / dx -> dy / dx
    # (B, D_x) -> (B, 1, D_x) -> (B, D_y, D_x)
    jac = [torch.autograd.grad(y[:, i], x, 
                               grad_outputs=vector, 
                               retain_graph=True,
                               create_graph=True)[0].view(batch, -1)
                for i in range(y_size)]
    jac = torch.stack(jac, dim=1)
    
    return jac
```

**Comment by dashesy** [comment]
Currently I use `torch.split` to split the channels (to different sizes) and then apply softmax separately on each tensor. It would be nice to avoid splitting and use the Dense Packed data layout on a tensor (something like `view` that does not copy data). I would need something like `torch.view_as_nested()` that could then be passed to a single `softmax` for reduction.

Currently this is how to compute softmax on a jagged tensor

```python

a = torch.arange(15).reshape(3,5).float()
b = torch.cat([b.softmax(0) for b in a.split((2,3), dim=1)], 1)

```

Would be nice to have 

```python

a = torch.arange(15).reshape(3,5).float()
b = a.view_as_nested((2,3), dim=1).softmax(0).contiguous()

```

The last `contiguous` should be a noop in this case that only changes the view

**Comment by cpuhrsch** [comment]
@dashesy  - Thanks for the comment!

At this point I'd implement something like this via

```
a = torch.arange(15).reshape(3,5).float()
nt = torch.nested_tensor(a.split((2,3), dim=1)) # Create new, contiguous NestedTensor via copy
nt_ = nt.softmax(0)
```

**Comment by NumesSanguis** [comment]
I didn't read the whole discussion above, but when I was looking for how to deal with audio files of different lengths, I found `awkward-array`: https://github.com/scikit-hep/awkward-array

It seems that it's basically numpy, but with support for `JaggedArrays`.
There seems to support for "awkd" files, HDF5, Parquet, or ROOT type storages, which would help with loading jagged arrays.

Also, it seems to have support for Tables, which could result in named arrays. That would combine this effort and NamedTensors (https://pytorch.org/docs/stable/named_tensor.html).

With `1.0.0` they plan to switch to C++ and have 2 implementations per algorithm, one for the CPU and 1 for the GPU. Would seem to play well when converting between Numpy and Tensors?:
https://github.com/scikit-hep/awkward-1.0

**Comment by vadimkantorov** [comment]
Nested tensor would also be useful for parallel reductions of a list of tensor without resorting to torch.stack first: https://github.com/pytorch/pytorch/issues/27522 which would save on memory, sometimes intermediate copmutations, and we could also separately track parallelization performance for 2/3-sized inputs (logadd is a frequent example)

**Comment by vadimkantorov** [comment]
A wild idea: `torch.cat` / `torch.stack` may be transformed to be NestedTensor constructors (if only tensor lists are considered) if some new argument is set.

**Comment by vadimkantorov** [comment]
Here is another in-the-wild case where a nested tensor could avoid torch.stack materialization: https://github.com/pytorch/pytorch/pull/32020/files

**Comment by kaiseryet** [comment]
If the wrapper does a deep copy of the tensors, it might not be efficient... am I right?

Also, what's the approx release date for `nestedtensor`? Waiting to use it asap.

**Comment by cpuhrsch** [comment]
Hey all, please see the main overview issue here: https://github.com/pytorch/pytorch/issues/25032 - I'll update this issue soon with plans for the next few months.

We also have a [separate repo](https://github.com/pytorch/nestedtensor) for NestedTensors. The project itself still is highly experimental, so don't rely on it for now.

**Comment by dbsxdbsx** [comment]
How is this feature now?

**Comment by cpuhrsch** [comment]
@dbsxdbsx - It landed [in core](https://pytorch.org/docs/master/nested.html).

**Comment by haotang1995** [comment]
Hi, may I know if it will support simple operators such as sum/mean? I am looking for [memory efficient way of implementing summation over a sequence of tensors](https://discuss.pytorch.org/t/memory-efficent-way-to-sum-over-a-sequence-of-tensors/160818). Thanks!

**Comment by mmcdermott** [comment]
I don't know if this will be relevant for anyone, but I just implemented [this repo](https://github.com/mmcdermott/nested_ragged_tensors) in response to some issues I was having where I wanted to store some nested ragged tensors to minimize memory and runtime costs when doing nested padding in data pre-processing. It does not implement core operations on these tensors, but it does permit efficient storage and slicing of nested 1D tensors which may be useful during data loading.
